numba/np/arraymath.py

```diff
@@ -3899,6 +3899,20 @@ def impl(a, dtype=None):
     return impl
 
 
+@overload(np.asfarray)
+def np_asfarray(a, dtype=types.float64):
+    if is_nonelike(dtype):
+        dt = types.float64
+    else:
+        dt = dtype.dtype
+        if not isinstance(dt, types.Float):
```

Comment:
numpy allows complex to be the output dtype of `asfarray` but it seems that this line would produce a float output, even when the user passes a complex dtype.